decayedlog.go

```diff
@@ -0,0 +1,419 @@
+
+// HashSharedSecret Sha-256 hashes the shared secret and returns the first
+// sharedHashSize bytes of the hash.
+func HashSharedSecret(sharedSecret *Hash256) HashPrefix {
+	// Sha256 hash of sharedSecret
+	h := sha256.New()
+	h.Write(sharedSecret[:])
+
+	var sharedHash HashPrefix
+
+	// Copy bytes to sharedHash
+	copy(sharedHash[:], h.Sum(nil))
+	return sharedHash
+}
+
+// Delete removes a <shared secret hash, CLTV> key-pair from the
+// sharedHashBucket.
+func (d *DecayedLog) Delete(hash []byte) error {
+	return d.db.Batch(func(tx *bolt.Tx) error {
+		sharedHashes := tx.Bucket(sharedHashBucket)
+		if sharedHashes == nil {
+			return ErrDecayedLogCorrupted
+		}
+
+		return sharedHashes.Delete(hash)
+	})
+}
+
+// Get retrieves the CLTV of a processed HTLC given the first 20 bytes of the
+// Sha-256 hash of the shared secret.
+func (d *DecayedLog) Get(hash []byte) (uint32, error) {
+	// math.MaxUint32 is returned when Get did not retrieve a value.
+	// This was chosen because it's not feasible for a CLTV to be this high.
+	var value uint32 = math.MaxUint32
+
+	err := d.db.View(func(tx *bolt.Tx) error {
+		// Grab the shared hash bucket which stores the mapping from
+		// truncated sha-256 hashes of shared secrets to CLTV's.
+		sharedHashes := tx.Bucket(sharedHashBucket)
+		if sharedHashes == nil {
+			return fmt.Errorf("sharedHashes is nil, could " +
+				"not retrieve CLTV value")
+		}
+
+		// Retrieve the bytes which represents the CLTV
+		valueBytes := sharedHashes.Get(hash)
+		if valueBytes == nil {
+			return nil
+		}
+
+		// The first 4 bytes represent the CLTV, store it in value.
+		value = uint32(binary.BigEndian.Uint32(valueBytes))
+
+		return nil
+	})
+	if err != nil {
+		return value, err
+	}
+
+	return value, nil
+}
+
+// Put stores a shared secret hash as the key and the CLTV as the value.
+func (d *DecayedLog) Put(hash *HashPrefix, cltv uint32) error {
+	// Optimisitically serialize the cltv value into the scratch buffer.
+	var scratch [4]byte
+	binary.BigEndian.PutUint32(scratch[:], cltv)
+
+	return d.db.Batch(func(tx *bolt.Tx) error {
+		sharedHashes := tx.Bucket(sharedHashBucket)
+		if sharedHashes == nil {
+			return ErrDecayedLogCorrupted
+		}
+
+		// Check to see if this hash prefix has been recorded before. If
+		// a value is found, this packet is being replayed.
+		valueBytes := sharedHashes.Get(hash[:])
+		if valueBytes != nil {
+			return ErrReplayedPacket
+		}
+
+		return sharedHashes.Put(hash[:], scratch[:])
+	})
+}
+
+func (d *DecayedLog) PutBatch(b *Batch) (*ReplaySet, error) {
+	// Since batched boltdb txns may be executed multiple times before
+	// succeeding, we will create a new replay set for each invocation to
+	// avoid any side-effects. If the txn is successful, this replay set
+	// will be merged with the replay set computed during batch construction
+	// to generate the complete replay set. If this batch was previously
+	// processed, the replay set will be deserialized from disk.
+	var replays *ReplaySet
```

Comment:
Your comment is correct, but I think this actually needs to be _within_ the scope of the closure function below, no?

Reply:
The intent was that `replays` be overwritten by repeated invocations of the closure, ensuring that the correct value is initialized if the write does succeed. Moving it inside would also make it out of scope to its assignment to `b.replaySet` at the end of the method